### PR TITLE
Dm 1284 Log Error when PV units change

### PR DESCRIPTION
**Pull request by rerpha**
### Description of work

Logging an error when a PVs unit changes when it has been set.

### Issue

DM-1284

### Acceptance Criteria

Changed EPICS monitor request to populate the pv structure with display.units and unpopulating alarms as they are seemingly not used. 

### Unit Tests

None. 

### Other

System test added for this. 

---

## Code Review (To be filled in by the reviewer only)

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [ ] Recommend for group code review

Also, nominate it on the code_review Slack channel (does someone want to automate this?).

### CODE REVIEW

File: src/ConversionWorker.cpp [hunk 1]
```diff
@@ -36,13 +36,14 @@ int ConversionWorker::run() {
       auto qf = queue.MAX_SUBQUEUE_SIZE - qs;
       scheduler->fill(queue, qf, id);
     }
+    std::string Units = "";
```
Comment (matthew-d-jones) [inline src/ConversionWorker.cpp]: The `=""` is superfluous

File: src/schemas/f142/f142.cpp [hunk 2]
```diff
@@ -395,6 +415,9 @@ class Converter : public FlatBufferCreator {
 
   RangeSet<uint64_t> seqs;
   Statistics Stats;
+
+private:
+  std::string CachedUnits = "";
```
Comment (matthew-d-jones) [inline src/schemas/f142/f142.cpp]: and again here, `=""` is not needed.

File: src/FlatBufferCreator.h [hunk 3]
```diff
@@ -13,7 +13,7 @@ struct EpicsPVUpdate;
 class FlatBufferCreator {
 public:
   virtual ~FlatBufferCreator() = default;
-  virtual std::unique_ptr<FlatbufferMessage>
+  virtual std::unique_ptr<FlatBufs::FlatbufferMessage>
```
Comment (matthew-d-jones) [inline src/FlatBufferCreator.h]: `FlatBufferCreator` is in the `FlatBufs` namespace, so don't need `FlatBufs::` on `FlatbufferMessage`

File: src/schemas/f142/f142.cpp [hunk 4]
```diff
@@ -5,6 +5,7 @@
 #include "../../logger.h"
 #include "schemas/f142_logdata_generated.h"
```
Comment (matthew-d-jones) [inline src/schemas/f142/f142.cpp]: I assume this include can be removed

File: src/schemas/f142/f142.cpp [hunk 5]
```diff
@@ -362,12 +363,25 @@ class Converter : public FlatBufferCreator {
     // this is the field type ID string: up.pvstr->getStructure()->getID()
     auto PVName = Builder->CreateString(PVUpdate.channel);
     auto Value = makeValue(*Builder, PVStructure, true, Stats);
-
     LogDataBuilder LogDataBuilder(*Builder);
     LogDataBuilder.add_source_name(PVName);
     LogDataBuilder.add_value_type(Value.Type);
     LogDataBuilder.add_value(Value.Offset);
 
+    if (auto PVDisplay =
```
Comment (matthew-d-jones) [inline src/schemas/f142/f142.cpp]: Could this code block be extracted to a separate method?

Comment (mchorazak) [reply src/schemas/f142/f142.cpp]: It's now separated into two methods - one for recording timestamp and one for checking units.

**Comment by rerpha** [comment]
The system test for this is not quite working on jenkins just yet. I think it is due to the log file only closing/flushing when the forwarder exits and in this case the container isn't shutting down quickly enough. maybe adding larger sleep after sending the exit command will help? if not, we'll have to work around this by shutting the container itself down.

File: src/schemas/f142/f142.cpp [hunk 6]
```diff
@@ -362,12 +362,30 @@ class Converter : public FlatBufferCreator {
...
+
+    FinishLogDataBuffer(*Builder, LogDataBuilder.Finish());
+    return FlatbufferMessage;
+  }
+
+  std::map<std::string, double> getStats() override {
+    return {{"ranges_n", seqs.size()}};
+  }
+
+  RangeSet<uint64_t> seqs;
```
[the remark this answers is not in the document]
Comment (mattclarke) [reply src/schemas/f142/f142.cpp]: seqs and Stats I mean. And can you change it to Seqs?

**Pull request merged**